okhttp/src/test/java/okhttp3/EventListenerTest.java

```diff
@@ -943,56 +948,77 @@ private void responseBodyFail(Protocol expectedProtocol) throws IOException {
         "ResponseBodyEnd", "ConnectionReleased", "CallEnd");
   }
 
-  @Test public void requestBodyFailHttp1OverHttps() throws IOException {
+  @Test public void requestBodyFailHttp1OverHttps() {
     enableTlsWithTunnel(false);
     server.setProtocols(asList(Protocol.HTTP_1_1));
-    requestBodyFail();
+
+    requestBodyFail(Protocol.HTTP_1_1);
   }
 
-  @Test public void requestBodyFailHttp2OverHttps() throws IOException {
+  @Test public void requestBodyFailHttp2OverHttps() {
     enableTlsWithTunnel(false);
     server.setProtocols(asList(Protocol.HTTP_2, Protocol.HTTP_1_1));
-    requestBodyFail();
+
+    requestBodyFail(Protocol.HTTP_2);
   }
 
-  @Test public void requestBodyFailHttp() throws IOException {
-    requestBodyFail();
+  @Test public void requestBodyFailHttp() {
+    requestBodyFail(null);
   }
 
-  private void requestBodyFail() {
-    // Stream a 256 MiB body so the disconnect will happen before the server has read everything.
-    RequestBody requestBody = new RequestBody() {
+  private void requestBodyFail(@Nullable Protocol requiredProtocol) {
+    CountDownLatch latch = new CountDownLatch(1);
+
+    try {
+      server.enqueue(new MockResponse()
+          .setSocketPolicy(SocketPolicy.DISCONNECT_DURING_REQUEST_BODY));
+
+      Call call = client.newCall(new Request.Builder()
+          .url(server.url("/"))
+          .post(nonCompletingRequest(latch))
+          .build());
+      try {
+        call.execute();
+        fail();
+      } catch (IOException expected) {
+      }
+
+      if (requiredProtocol != null) {
+        ConnectionAcquired connectionAcquired = listener.removeUpToEvent(ConnectionAcquired.class);
+        assertThat(connectionAcquired.connection.protocol()).isEqualTo(requiredProtocol);
+      }
+
+      CallFailed callFailed = listener.removeUpToEvent(CallFailed.class);
+      assertThat(callFailed.ioe).isNotNull();
+    } finally {
+      latch.countDown();
+    }
+  }
+
+  @NotNull private RequestBody nonCompletingRequest(CountDownLatch latch) {
+    return new RequestBody() {
       @Override public MediaType contentType() {
         return MediaType.get("text/plain");
       }
 
       @Override public long contentLength() {
-        return 1024 * 1024 * 256;
+        return 1024 * 1024 * 4;
       }
 
       @Override public void writeTo(BufferedSink sink) throws IOException {
-        for (int i = 0; i < 1024; i++) {
-          sink.write(new byte[1024 * 256]);
-          sink.flush();
+        sink.write(new byte[1024 * 1024]);
+        sink.flush();
+        sink.write(new byte[1024 * 1024]);
+        sink.flush();
+        sink.write(new byte[1024 * 1024]);
```

Comment:
To cause the client to fail writing the request body, don't you have to ensure that the client tries to write at least once *after* the server disconnects? I don't see what guarantees that this is always the case here.